File: app/src/main/java/com/soya/launcher/ui/fragment/AppsFragment.java
package com.soya.launcher.ui.fragment;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.ApplicationInfo;
import android.content.pm.ResolveInfo;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.leanback.widget.ArrayObjectAdapter;
import androidx.leanback.widget.FocusHighlight;
import androidx.leanback.widget.FocusHighlightHelper;
import androidx.leanback.widget.ItemBridgeAdapter;
import androidx.leanback.widget.VerticalGridView;

import com.soya.launcher.App;
import com.soya.launcher.R;
import com.soya.launcher.adapter.AppListAdapter;
import com.soya.launcher.ui.dialog.AppDialog;
import com.soya.launcher.utils.AndroidSystem;

import java.util.List;

public class AppsFragment extends AbsFragment{

    public static AppsFragment newInstance() {
        
        Bundle args = new Bundle();
        
        AppsFragment fragment = new AppsFragment();
        fragment.setArguments(args);
        return fragment;
    }

    private VerticalGridView mContentGrid;
    private TextView mTitleView;

    private InnerReceiver receiver;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        receiver = new InnerReceiver();
        IntentFilter filter = new IntentFilter();
        filter.addAction(Intent.ACTION_PACKAGE_ADDED);
        filter.addAction(Intent.ACTION_PACKAGE_REMOVED);
        filter.addAction(Intent.ACTION_PACKAGE_REPLACED);
        filter.addDataScheme("package");
        getActivity().registerReceiver(receiver, filter);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        getActivity().unregisterReceiver(receiver);
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_apps;
    }

    @Override
    protected void init(View view, LayoutInflater inflater) {
        super.init(view, inflater);
        mContentGrid = view.findViewById(R.id.content);
        mTitleView = view.findViewById(R.id.title);

        mTitleView.setText(getString(R.string.apps));
    }

    @Override
    protected void initBind(View view, LayoutInflater inflater) {
        super.initBind(view, inflater);
        fillApps();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        requestFocus(mContentGrid);
    }

    @Override
    protected int getWallpaperView() {
        return R.id.wallpaper;
    }

    private void fillApps(){
        setContent(AndroidSystem.getUserApps(getActivity()));
    }

    private void setContent(List<ApplicationInfo> list){
        ArrayObjectAdapter arrayObjectAdapter = new ArrayObjectAdapter(new AppListAdapter(getActivity(), getLayoutInflater(), R.layout.holder_app_2, newAppListCallback()));
        ItemBridgeAdapter itemBridgeAdapter = new ItemBridgeAdapter(arrayObjectAdapter);
        FocusHighlightHelper.setupBrowseItemFocusHighlight(itemBridgeAdapter, FocusHighlight.ZOOM_FACTOR_MEDIUM, false);
        mContentGrid.setAdapter(itemBridgeAdapter);
        mContentGrid.setNumColumns(4);
        arrayObjectAdapter.addAll(0, list);
        requestFocus(mContentGrid);
    }

    private AppListAdapter.Callback newAppListCallback(){
        return new AppListAdapter.Callback() {

            @Override
            public void onSelect(boolean selected) {
            }

            @Override
            public void onClick(ApplicationInfo bean) {
                if (App.COMPANY == 1 || App.COMPANY == 2 || App.COMPANY == 3){
                    AndroidSystem.openPackageName(getActivity(), bean.packageName);
                }else {
                    AppDialog dialog = AppDialog.newInstance(bean);
                    dialog.setCallback(new AppDialog.Callback() {
                        @Override
                        public void onDelete() {
                            AndroidSystem.openApplicationDetials(getActivity(), bean.packageName);
                        }

                        @Override
                        public void onOpen() {
                            AndroidSystem.openPackageName(getActivity(), bean.packageName);
                        }
                    });
                    dialog.show(getChildFragmentManager(), AppDialog.TAG);
                }
            }
        };
    }

    public class InnerReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {
            switch (intent.getAction()){
                case Intent.ACTION_PACKAGE_ADDED:
                case Intent.ACTION_PACKAGE_REMOVED:
                case Intent.ACTION_PACKAGE_REPLACED:
                    fillApps();
                    break;
            }
        }
    }
}

File: app/src/main/java/com/soya/launcher/ui/fragment/AppsFragment.kt
package com.soya.launcher.ui.fragment

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.ApplicationInfo
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.animation.AnimationUtils
import android.widget.ImageView
import android.widget.TextView
import androidx.core.os.HandlerCompat.postDelayed
import androidx.leanback.widget.VerticalGridView
import androidx.lifecycle.lifecycleScope
import com.drake.brv.utils.grid
import com.drake.brv.utils.models
import com.drake.brv.utils.mutable
import com.drake.brv.utils.setup
import com.shudong.lib_base.base.BaseViewModel
import com.shudong.lib_base.ext.clickNoRepeat
import com.shudong.lib_base.ext.no
import com.soya.launcher.BaseWallPaperFragment
import com.soya.launcher.R
import com.soya.launcher.config.Config
import com.soya.launcher.databinding.FragmentAppsBinding
import com.soya.launcher.ui.dialog.AppDialog
import com.soya.launcher.utils.AndroidSystem
import com.soya.launcher.utils.isSysApp
import com.soya.launcher.view.AppLayout
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class AppsFragment : BaseWallPaperFragment<FragmentAppsBinding,BaseViewModel>() {

    private var receiver: InnerReceiver? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        lifecycleScope.launch {
            receiver = InnerReceiver()
            val filter = IntentFilter()
            filter.addAction(Intent.ACTION_PACKAGE_ADDED)
            filter.addAction(Intent.ACTION_PACKAGE_REMOVED)
            filter.addAction(Intent.ACTION_PACKAGE_REPLACED)
            filter.addDataScheme("package")
            activity!!.registerReceiver(receiver, filter)
        }


    }

    override fun onDestroy() {
        super.onDestroy()
        activity!!.unregisterReceiver(receiver)
    }


    override fun initView() {
        mBind.content.grid(4).setup {
            addType<ApplicationInfo>(R.layout.item_apps)
            onBind {

                val bean = _data as ApplicationInfo
                val mIV = findView<ImageView>(R.id.image)
                val mTitle = findView<TextView>(R.id.title)
                val mIVSmall = findView<ImageView>(R.id.image_small)
                val mAppLayout = findView<AppLayout>(R.id.fl_focus)

                val pm = context.packageManager
                val banner: Drawable? = null
                if (banner != null) {
                    mIV.setImageDrawable(banner)
                } else {
                    mIVSmall.setImageDrawable(bean.loadIcon(pm))
                }
                if (Config.COMPANY == 4 && bean.packageName == "com.mediatek.wwtv.tvcenter") {
                    mTitle.text = "HDMI"
                } else {
                    mTitle.text = bean.loadLabel(pm)
                }
                mIV.visibility = if (banner == null) View.GONE else View.VISIBLE
                mIVSmall.visibility = if (banner == null) View.VISIBLE else View.GONE

                mAppLayout.onFocusChangeListener = View.OnFocusChangeListener { v, hasFocus ->
                    itemView.isSelected = hasFocus
                    val animation = AnimationUtils.loadAnimation(
                        context, if (hasFocus) R.anim.zoom_in_middle else R.anim.zoom_out_middle
                    )
                    itemView.startAnimation(animation)
                    animation.fillAfter = true
                }

                mAppLayout.clickNoRepeat {
                    AndroidSystem.openPackageName(requireContext(), bean.packageName)
                }

                mAppLayout.setListener(AppLayout.EventListener { keyCode, event ->
                    if ( event?.keyCode == KeyEvent.KEYCODE_MENU) {
                        isSysApp(bean.packageName).no {
                            //点击了菜单
                            menu(bean)
                        }
                        return@EventListener false
                    }
                    true
                })


            }
        }.models = arrayListOf()

        mBind.layout.title.text = getString(R.string.apps)
    }



    private fun fillApps() {
        lifecycleScope.launch {
            var list:List<ApplicationInfo>
            withContext(Dispatchers.IO){
                 list = AndroidSystem.getUserApps(requireContext())
            }
            withContext(Dispatchers.Main){
                setContent(list)
            }
        }
    }

    private fun setContent(list: List<ApplicationInfo>) {
       // val filteredList = list.toMutableList().let { product.filterRepeatApps(it) } ?:list
        if(list.size!=mBind.content.mutable.size){
            mBind.content.models = list
            mBind.content.apply {
                postDelayed({
                    requestFocus()
                    layoutManager?.findViewByPosition(0)?.requestFocus()
                },800)
            }
        }

        //requestFocus(mContentGrid)
    }


    private fun menu(bean: ApplicationInfo) {
        val dialog = AppDialog.newInstance(bean)
        dialog.setCallback (object :AppDialog.Callback{
            override fun onOpen() {
                AndroidSystem.openPackageName(
                    requireContext(), bean.packageName
                )
            }

        })
        dialog.show(childFragmentManager, AppDialog.TAG)
    }

    inner class InnerReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            when (intent.action) {
                Intent.ACTION_PACKAGE_ADDED, Intent.ACTION_PACKAGE_REMOVED, Intent.ACTION_PACKAGE_REPLACED -> fillApps()
            }
        }
    }

    override fun onResume() {
        super.onResume()
        fillApps()

    }

    companion object {
        fun newInstance(): AppsFragment {
            val args = Bundle()

            val fragment = AppsFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
